### PR TITLE
Fix token performance telemetry and stats reporting (Fixes #1805)

### LINKED ISSUE #1805: TPM counts only output tokens; TTFT never populated; /stats missing performance metrics
### What happened?

The TPM (tokens per minute) metric was intended to measure total throughput from input to output — how much work the model does per minute of pure model+network time, excluding approvals, tool execution, and user interaction. Instead, it only counts **output tokens** in the numerator.

Additionally, several related performance metrics are broken or missing, making it hard for users to compare providers/models for coding workloads.

### Problems

**1. TPM numerator is output-only**

Both callsites in LoggingProviderWrapper.ts pass only output_token_count to recordCompletion:

    const outputTokens = tokenCounts.output_token_count;
    this.performanceTracker.recordCompletion(duration, null, outputTokens, 0);

The time span (denominator) correctly measures pure model+network time, but the token count (numerator) should be input_token_count + output_token_count to reflect total throughput.

**2. TTFT (time to first token) is never populated**

ProviderPerformanceTracker.recordCompletion accepts a timeToFirstToken parameter and the metrics struct has the field, but both production callsites always pass null. The stream loop in processStreamForMetrics and logResponseStream already iterates chunks — it just needs to record performance.now() - startTime on the first yielded chunk.

**3. chunkCount always passed as 0**

Both callsites pass 0 for chunkCount, so metrics.chunksReceived is always wrong. The stream loops already process chunks and could easily count them.

**4. tokensPerSecond overwrites on every call**

    this.metrics.tokensPerSecond = tokenCount / (totalTime / 1000);

This is just the last request rate, not a rolling or session average. Misleading if inspected.

**5. /stats Performance section is missing key metrics**

The /stats display shows Wall Time, Agent Active, API Time %, and Tool Time %, but does NOT show:
- TPM (the status bar metric)
- TTFT
- Tokens/sec (output generation rate)
- Total throughput (input+output tokens / model time)

The Model Usage table shows token counts but no rate/speed information per model.

### What did you expect to happen?

- **TPM** should reflect total tokens (input + output) processed per minute of model+network time
- **TTFT** should be measured and displayed — this is the dominant factor in perceived speed with large contexts
- **chunkCount** should be tracked correctly
- **tokensPerSecond** should be meaningful (rolling average or clearly labeled as last-request)
- **/stats Performance section** should surface TPM, TTFT, and throughput metrics so users can compare providers

### Fix locations

- packages/core/src/providers/LoggingProviderWrapper.ts — both recordCompletion callsites (lines ~776 and ~919): pass input + output tokens, measure TTFT from first chunk, count chunks
- packages/core/src/providers/logging/ProviderPerformanceTracker.ts — fix tokensPerSecond to be a rolling average or rename it
- packages/cli/src/ui/components/StatsDisplay.tsx — add TPM, TTFT, throughput to Performance section
- packages/cli/src/ui/components/Footer.tsx — TPM display will automatically reflect corrected values

### Anything else we need to know?

The time measurement (denominator) IS correct for the intended goal — duration in processStreamForMetrics and logResponseStream is performance.now() - startTime spanning request-start to last-chunk, which is pure model+network time with no approval/tool/user delays mixed in.

This is a follow-up to #1764 which fixed TPM being 0.0 for stream calls. That fix correctly moved recordCompletion outside the if (latestTokenUsage) guard and added tiktoken-based output estimation, but the underlying issue of output-only counting was pre-existing.

### PR DESCRIPTION
**Pull request by acoliver**
## Summary
- fixes throughput tracking so tokens per minute uses input plus output tokens instead of output-only counts
- captures time to first token from the first token-bearing streamed chunk and propagates real chunk counts
- converts tokens per second from last-request overwrite behavior to cumulative session throughput
- wires new performance fields through telemetry and token metrics tracking for UI consumers
- adds Throughput, TTFT, and Output Rate rows to the stats performance section with conditional display rules

## Validation
- npm run test
- NODE_OPTIONS=--max-old-space-size=8192 npm run lint
- npm run typecheck
- npm run format
- npm run build
- node scripts/start.js --profile-load synthetic "write me a haiku and nothing else"

## Notes
- includes test updates across core provider telemetry, performance tracker, ui telemetry pipeline, and stats display snapshots

### CODE REVIEW

File: packages/cli/src/ui/components/StatsDisplay.tsx [hunk 1]
```diff
@@ -359,6 +359,30 @@ export const StatsDisplay: React.FC<StatsDisplayProps> = ({
...
+            </Text>
+          </SubStatRow>
+        )}
+        {metrics.tokenTracking.tokensPerSecond > 0 && (
+          <SubStatRow title="Output Rate:">
+            <Text color={theme.text.primary}>
+              {metrics.tokenTracking.tokensPerSecond.toFixed(2)} tok/s
+            </Text>
+          </SubStatRow>
+        )}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/ui/components/StatsDisplay.tsx]: Addressed in commit 559d60843. I added finite value guards for throughput, TTFT, and output rate rendering in StatsDisplay so non-finite values are skipped. I also added a regression test that verifies NaN and Infinity do not render and updated the snapshot accordingly.

File: packages/core/src/providers/logging/ProviderPerformanceTracker.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ export class ProviderPerformanceTracker {
     completionTimestamp: number;
     tokenCount: number;
   }>;
+  private totalGenerationTimeMs = 0;
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/core/src/providers/logging/ProviderPerformanceTracker.ts:22]: Addressed in commit 559d60843. I paired TPS token accumulation with positive elapsed time by introducing a measured-time token accumulator that only increments when totalTime is greater than 0. tokensPerSecond is now computed from matched token and time pairs, and reset clears this accumulator. I also added a test that asserts a zero-time completion does not inflate the cumulative rate.

File: packages/core/src/providers/LoggingProviderWrapper.ts [hunk 3]
```diff
@@ -864,11 +886,32 @@ export class LoggingProviderWrapper implements IProvider {
...
+        if (block.type === 'text') {
+          return typeof block.text === 'string' && block.text.length > 0;
+        }
+        return block.type === 'thinking' || block.type === 'code';
+      });
+    }
+
+    const extractedContent = this.extractSimpleContent(chunk);
+    return extractedContent.length > 0;
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/core/src/providers/LoggingProviderWrapper.ts:922]: Addressed in commit 559d60843. I updated token-bearing detection to include tool_call blocks in hasTokenBearingOutput so TTFT is captured for tool-call-only streaming responses. I also added a provider test that exercises a tool_call-only stream and verifies TTFT is populated.

File: packages/cli/src/ui/components/StatsDisplay.test.tsx [hunk 4]
```diff
@@ -622,4 +655,154 @@ describe('<StatsDisplay />', () => {
...
+          tokensPerSecond: Number.NEGATIVE_INFINITY,
+        },
+      });
+
+      const { lastFrame } = renderWithMockedStats(metrics);
+      const output = lastFrame();
+
+      expect(output).not.toContain('Throughput:');
+      expect(output).not.toContain('TTFT:');
+      expect(output).not.toContain('Output Rate:');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/ui/components/StatsDisplay.test.tsx]: Addressed in commit e2859798a. I updated all stale negative assertions in StatsDisplay.test.tsx to use the current label Token Rate: instead of Output Rate: (the three occurrences flagged in this thread). This makes the tests verify the intended hidden-row behavior against the current UI label.

File: packages/cli/src/ui/components/StatsDisplay.test.tsx [hunk 5]
```diff
@@ -622,4 +655,155 @@ describe('<StatsDisplay />', () => {
...
+          tokensPerSecond: Number.NEGATIVE_INFINITY,
+        },
+      });
+
+      const { lastFrame } = renderWithMockedStats(metrics);
+      const output = lastFrame();
+
+      expect(output).not.toContain('Throughput:');
+      expect(output).not.toContain('TTFT:');
+      expect(output).not.toContain('Token Rate:');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/ui/components/StatsDisplay.test.tsx]: Fixed in the current branch. I updated all three stale negative assertions in StatsDisplay.test.tsx to the current labels: `TTFT (last):` and `Token Rate (avg):` (keeping Throughput unchanged, since the UI title is still `Throughput:`). This restores the hidden-state guard behavior for the active labels.

File: packages/core/src/providers/logging/ProviderPerformanceTracker.test.ts [hunk 6]
```diff
@@ -203,4 +215,137 @@ describe('ProviderPerformanceTracker', () => {
       'Provider: test-provider, Requests: 1, Avg Latency: 1000.00ms, Tokens/sec: 500.00, Error Rate: 50.0%',
     );
   });
+
+  describe('Issue #1805: TPM numerator uses total tokens (input + output)', () => {
+    it('should accumulate totalTokens as input + output for each completion', () => {
+      vi.useRealTimers();
+      const tracker = new ProviderPerformanceTracker('test-provider');
+      vi.useFakeTimers();
+      vi.setSystemTime(new Date('2025-01-01T00:00:00Z').getTime());
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/core/src/providers/logging/ProviderPerformanceTracker.test.ts]: Good call. I simplified timer handling by centralizing cleanup with `afterEach(() => vi.useRealTimers())` and removed the redundant per-test `vi.useRealTimers()` toggles around the issue-1805 timer tests. This keeps intent clear and avoids defensive switching in each test block.

File: packages/core/src/providers/logging/ProviderPerformanceTracker.ts [hunk 7]
```diff
@@ -19,6 +19,8 @@ export class ProviderPerformanceTracker {
...
+  private totalTokensWithMeasuredTime = 0;
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/core/src/providers/logging/ProviderPerformanceTracker.ts:23]: Fixed in this branch. I reset tracker session accumulators at the same boundary as wrapper state reset by adding `this.performanceTracker.reset()` in `LoggingProviderWrapper.clearState()`. I also added a regression test in LoggingProviderWrapper.apiTelemetry.test.ts that verifies metrics are non-zero after a run, then return to baseline after `clearState?.()`.

File: packages/core/src/providers/LoggingProviderWrapper.ts [hunk 8]
```diff
@@ -914,9 +968,15 @@ export class LoggingProviderWrapper implements IProvider {
...
-      this.performanceTracker.recordCompletion(duration, null, outputTokens, 0);
+      // Record performance metrics (TPM tracks total tokens: input+output)
+      const perfTotalTokens =
+        tokenCounts.input_token_count + tokenCounts.output_token_count;
+      this.performanceTracker.recordCompletion(
+        duration,
+        timeToFirstToken ?? null,
+        perfTotalTokens,
+        chunkCount ?? 0,
+      );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/core/src/providers/LoggingProviderWrapper.ts]: Fixed in the current branch. In `LoggingProviderWrapper.logResponse(...)` I now branch by `success`: successful streams call `recordCompletion(...)`, failed streams call `recordError(...)` with duration/error/TTFT/chunk metadata. Added regression coverage in LoggingProviderWrapper.apiTelemetry.test.ts to verify failed logResponseStream paths increment errors and do not increase completion counters/tokens.
